src/cache.c

```diff
@@ -129,6 +132,7 @@ hsk_cache_insert(
 
   if (!hsk_cache_insert_data(c, req->name, req->type, wire, wire_len)) {
     hsk_cache_log(c, "could not insert cache\n");
+    free(wire);
```

Comment:
👍  had to double check that `hsk_dns_msg_encode` would not return `false` if the malloc was successful (otherwise I'd expect a `free()` in the previous block as well).

Reply:
yeah, this makes code pretty unreadable.
That's the reason I mention to refactor so the callers should allocate the wire.

Reply:
But because this malloced data later is stored as reference in the cache map, so we can't free it
also make it harder to follow the logic.